windows/README_MINGW

```diff
@@ -0,0 +1,218 @@
+
+Or just download from a direct link (GCC 8.1.0):
+
+wget https://downloads.sourceforge.net/project/mingw-w64/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/8.1.0/threads-posix/sjlj/x86_64-8.1.0-release-posix-sjlj-rt_v6-rev0.7z
+
+Once downloaded:
+
+p7zip -d x86_64-8.1.0-release-posix-sjlj-rt_v6-rev0.7z
+
+That will decompress to a folder called mingw64 in the working directory.
+Now replace the files from the previous mingw-w64-x86_64-toolchain package
+installation:
+
+mv /mingw64 /mingw64.pacman
+mv mingw64 /
+
+Make sure GCC still runs and it is the version we need:
+
+$ gcc --version
+gcc.exe (x86_64-posix-sjlj-rev0, Built by MinGW-W64 project) {version}
+
+
+
+Preparing JACK dependencies
+---------------------------
+
+There are prebuilt MinGW binaries for all the libraries that can be installed
+using the pacman package manager, but since we are using a compiler that is not
+the default version shipped by MinGW, it seems better idea to build from source
+to avoid any linker and runtime issues. A good technical explanation and/or
+evidence for this statement is currently missing from this guide.
+
+Fortunately there are PKGBUILD files for doing so together with a nice guide at
+https://github.com/msys2/MINGW-packages
+
+git clone https://github.com/msys2/MINGW-packages.git
+
+cd MINGW-packages/mingw-w64-{libname}-git
+MINGW_INSTALLS=mingw64 makepkg-mingw -sLf
+pacman -U mingw-w64-{libname}-git-{suffix}.pkg.tar.xz
+
+Repeat procedure for each library replacing {libname} with the appropriate name
+and {suffix} with whatever the above process created.
+
+db
+libsndfile
+libsamplerate
+
+Some libraries like libsndfile and libsamplerate will ask for installing extra
+dependencies, it is ok to do so. The following need to be built in order:
+
+libtre
+libsystre
+
+libsystre is a wrapper around libtre that allows including <regex.h> later.
+Before building it adjust the includes path, it can be unset afterwards:
+
+export C_INCLUDE_PATH=/mingw64/include
+
+Finally for portaudio the Steinberg ASIO SDK should be manually downloaded.
+It can be found at https://www.steinberg.net/en/company/developers.html. The waf
+script will later check if the SDK is present at /opt/asiosdk
+
+wget https://www.steinberg.net/asiosdk -O /tmp/asiosdk.zip
+unzip /tmp/asiosdk.zip -d /tmp
+mv /tmp/asiosdk_{version}_{date} /opt/asiosdk
+
+The PKGBUILD description file needs to be patched so configure is called with
+the necessary flags for ASIO:
+
+--with-asiodir=/opt/asiosdk   <- new option     
+--with-winapi=wmme,directx,wasapi,vdmks,asio   <- append 'asio' to existing
+
+Currently the shared library version of portaudio fails to build, a further
+patch is needed so only the static version is generated. Comment out lines
+[61-75], [82-83] from PKGBUILD.
+
+
+
+Compiling JACK
+--------------
+
+- Clone repo
+
+git clone https://github.com/jackaudio/jack2
+
+- Build and install
+
+cd jack2
+./waf configure --prefix=/opt/jack && ./waf -p install
```

Comment:
The first `./waf` command fails here. From the config.log:
```
err: ../test.cpp:1:10: fatal error: opus/opus_custom.h: No such file or directory
 #include <opus/opus_custom.h>
          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
```
This happens nine times for the libraries/headers `opus`, `readline`, `systemd`, `db`, `alloca`, `poll`, `execinfo` and `regex`. I tried to build and install them from source as the ones above - is that the right direction?

Building `readline` from source fails with:
```
../../readline-8.0/colors.c: In function '_rl_print_color_indicator':
../../readline-8.0/colors.c:178:27: warning: implicit declaration of function 'S_ISLNK'; did you mean 'S_ISBLK'? [-Wimplicit-function-declaration]
   else if (linkok == 0 && S_ISLNK(mode) && _rl_color_indicator[C_ORPHAN].string != NULL)
                           ^~~~~~~
                           S_ISBLK
```
How did you fix this?

Reply:
Never happened to me. Let me build from scratch following the guide and see. Thanks.

Reply:
It works here, however I found three issues in the guide. Please let me know if the following helps:

1. If `pacman -Suy` ends with the following warning, you need to close the terminal AND run `pacman -Suy` again:

```
warning: terminate MSYS2 without returning to shell and check for updates again
warning: for example close your terminal window instead of calling exit
```

So please check there are no pending updates in your installation by running `pacman -Suy` again

2. This should be set before building any library:

`export C_INCLUDE_PATH=/mingw64/include`

3. Note the guide calls for cloning the official repository, which does not contain the wscript patches in this PR. Instead of cloning https://github.com/jackaudio/jack2 use fork  https://github.com/iamluciano/jack2, branch `target-mingw`.

I confirm your portaudio PKGBUILD patch works and allows building static and dynamic libraries.

Will update the guide soon, thanks for testing and sharing feedback.

Reply:
Also note no dependencies are needed beyond those currently indicated in the guide. Missing headers is a symptom of something else.

Reply:
There is a possibility the compiler toolchain installation fails in a not very explicit way, so I added details and an alternate solution for that step:

(Line 58 onwards)

That will decompress to a folder called `mingw64` in the working directory.
Now replace the files from the previous` mingw-w64-x86_64-toolchain` package
installation:

```
mv /mingw64 /mingw64.bak   -> backup original
rm /mingw64                -> see explanation below
mv mingw64 /               -> sjlj toolchain
```

The first step results in a file called `mingw64` file being created in `/`. The
file is not visible to the Windows File Explorer when inspecting `c:\msys64`, so
it is probably an artifact and safe to delete it. Otherwise the last `mv` step
will fail with:

`mv: cannot overwrite non-directory '/mingw64' with directory 'mingw64'`

An alternate solution consists in closing the MSYS2 window and replicating the
above procedure using the File Explorer. The following assumes the toolchain 7z
file was decompressed to the home directory:

Rename `c:\msys64\mingw64` to `c:\msys64\mingw64.bak`
Move `c:\msys64\home\{username}\mingw64` to `c:\msys64`

Reply:
>     3. Note the guide calls for cloning the official repository, which does not contain the wscript patches in this PR. Instead of cloning https://github.com/jackaudio/jack2 use fork  https://github.com/iamluciano/jack2, branch `target-mingw`.
That was the problem. Could have thought of that. So now the first command works.  

However, calling `./waf -p install` fails and I'm not sure where to look for an error message. I tried running this in verbose and it seems that the error is coming from `collect2.exe` (line 27 in the pastebin link), but I'm not sure if there is any more info. Here is a pastebin of the terminal output: https://pastebin.com/nBFEb89r  
Any idea where to dig for more info or what is causing this?

Reply:
Looks like the portaudio library is not included during link. Can you double check portaudio is building fine and the resulting .xz package is properly installed?

```
cd ~/MINGW-packages/mingw-w64-portaudio
pacman -U *.xz
```

Then check if waf configure is seeing portaudio:

`Checking for 'portaudio-2.0 >= 19'       : yes`

Reply:
Strange, I'm sure I already did that. Re-doing these steps fixed it. Nevermind!

Reply:
Strange, I'm sure I already did that. Re-doing these steps fixed it. Nevermind!